src/attachment.rs

```diff
@@ -0,0 +1,338 @@
+/// within `iterator`, passing `context`, and returning any non-zero value immediately (breaking iteration).
+pub type z_attachment_iter_driver_t = Option<
+    extern "C" fn(
+        iterator: *const c_void,
+        loop_body: z_attachment_iter_body_t,
+        context: *mut c_void,
+    ) -> i8,
+>;
+
+/// A iteration based map of byte slice to byte slice.
+///
+/// `iteration_driver == NULL` marks the gravestone value, as this type is often optional.
+/// Users are encouraged to use `z_attachment_null` and `z_attachment_check` to interact.
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct z_attachment_t {
+    pub data: *const c_void,
+    pub iteration_driver: z_attachment_iter_driver_t,
+}
+
+/// Returns the gravestone value for `z_attachment_t`.
+#[no_mangle]
+pub extern "C" fn z_attachment_check(this: &z_attachment_t) -> bool {
+    this.iteration_driver.is_some()
+}
+
+/// Returns the gravestone value for `z_attachment_t`.
+#[no_mangle]
+pub extern "C" fn z_attachment_null() -> z_attachment_t {
+    z_attachment_t {
+        data: core::ptr::null_mut(),
+        iteration_driver: None,
+    }
+}
+
+/// Iterate over `this`'s key-value pairs, breaking if `body` returns a non-zero
+/// value for a key-value pair, and returning the latest return value.
+///
+/// `context` is passed to `body` to allow stateful closures.
+///
+/// This function takes no ownership whatsoever.
+#[no_mangle]
+pub extern "C" fn z_attachment_iterate(
+    this: z_attachment_t,
+    body: z_attachment_iter_body_t,
+    context: *mut c_void,
+) -> i8 {
+    (this.iteration_driver.unwrap())(this.data, body, context)
+}
+
+/// Returns the value associated with the key.
+#[no_mangle]
+pub extern "C" fn z_attachment_get(this: z_attachment_t, key: z_bytes_t) -> z_bytes_t {
+    struct attachment_get_iterator_context {
+        key: z_bytes_t,
+        value: z_bytes_t,
+    }
+
+    extern "C" fn attachment_get_iterator(
+        key: z_bytes_t,
+        value: z_bytes_t,
+        context: *mut c_void,
+    ) -> i8 {
+        unsafe {
+            let context = &mut *(context as *mut attachment_get_iterator_context);
+            if context.key.as_slice() == key.as_slice() {
+                context.value = value;
+                1
+            } else {
+                0
+            }
+        }
+    }
+
+    let mut context = attachment_get_iterator_context {
+        key,
+        value: z_bytes_null(),
+    };
+
+    if this.iteration_driver.map_or(false, |iteration_driver| {
+        (iteration_driver)(
+            this.data,
+            attachment_get_iterator,
+            &mut context as *mut _ as *mut c_void,
+        ) != 0
+    }) {
+        context.value
+    } else {
+        z_bytes_null()
+    }
+}
+
+/// A map of maybe-owned vector of bytes to owned vector of bytes.
+///
+/// In Zenoh C, this map is backed by Rust's standard HashMap, with a DoS-resistant hasher
+#[repr(C)]
+pub struct z_owned_bytes_map_t {
```

Comment:
There is no guarantee that on different platforms size and alignment of this structure will match `HashMap<....>` will remain the same.
There is `impl_guarded_transmute` macro for solve it: https://github.com/eclipse-zenoh/zenoh-c/blob/d55da26aa8e53c1f83c05e3c2120afc5267fcd0e/src/publisher.rs#L69C9-L69C9

Reply:
Correct, do you mind doing the impl @milyin ?